src/java/src/main/java/com/dnanexus/DXSearch.java

```diff
@@ -989,14 +1013,111 @@ private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+        private final String classConstraint;
+        private final DXEnvironment env;
+        private final FindDataObjectsResponse response;
+
+        /**
+         * Iterator implementation for findDataObjects results page items.
+         */
+        @VisibleForTesting
+        private class ResultIterator implements Iterator<T> {
+
+            int lastElementIndex;
+
+            private ResultIterator() {
+                this.lastElementIndex = -1;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return lastElementIndex < (response.results.size() - 1);
+            }
+
+            @Override
+            public T next() {
+                return getDataObjectInstanceFromResult(response.results.get(++lastElementIndex));
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            @SuppressWarnings("unchecked")
+            private T getDataObjectInstanceFromResult(FindDataObjectsResponse.Entry e) {
+                DXDataObject dataObject = null;
+                DXContainer container = DXContainer.getInstance(e.project);
+                if (e.describe != null) {
+                    dataObject =
+                            DXDataObject.getInstanceWithCachedDescribe(e.id, container, env,
+                                    e.describe);
+                } else {
+                    dataObject = DXDataObject.getInstanceWithEnvironment(e.id, container, env);
+                }
+
+                if (classConstraint != null) {
+                    if (!dataObject.getId().startsWith(classConstraint + "-")) {
+                        throw new IllegalStateException("Expected all results to be of type "
+                                + classConstraint + " but received an object with ID "
+                                + dataObject.getId());
+                    }
+                }
+                // This is an unchecked cast, but the callers of this class
+                // should have set T appropriately so that it agrees with
+                // the class constraint (if any). If something goes wrong
+                // here, either that code is incorrect or the API server
+                // has returned incorrect results.
+                return (T) dataObject;
+            }
+        }
+
+        /**
+         * Initializes result set page object with the specified page size.
+         */
+        private SearchPage(FindDataObjectsRequest request, String classConstraint,
+                                      DXEnvironment env) {
+            this.classConstraint = classConstraint;
+            this.env = env;
+            this.response = DXAPI.systemFindDataObjects(request, FindDataObjectsResponse.class, env);
+        }
+
+        /**
+         * Returns an amount of items on findDataObjects results page
+         */
+        int size() {
```

Comment:
Actually I wanted it to be a part of public API, so I made it **public**. And unit-test checks that as well...